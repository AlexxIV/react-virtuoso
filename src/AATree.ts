interface NodeData<T> {
  key: number
  value: T
}

interface Range<T> {
  start: number
  end: number
  value: T
}

type FindCallback<T> = (value: T) => 1 | 0 | -1

export type NodeIterator<T> = IterableIterator<NodeData<T>>
export type RangeIterator<T> = IterableIterator<Range<T>>

class NilNode {
  public level = 0

  public rebalance(): this {
    return this
  }

  public adjust(): this {
    return this
  }

  public remove(): this {
    return this
  }

  public find(): void {
    return
  }

  public findWith(): void {
    return
  }

  public findMax(): number {
    return -Infinity
  }

  public insert<T>(key: number, value: T): NonNilNode<T> {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new NonNilNode<T>({ key, value, level: 1 })
  }

  public walkWithin<T>(): NodeData<T>[] {
    return []
  }

  public walk<T>(): NodeData<T>[] {
    return []
  }

  public ranges<T>(): Range<T>[] {
    return []
  }

  public rangesWithin<T>(): Range<T>[] {
    return []
  }

  public empty(): this is NilNode {
    return true
  }

  public isSingle(): true {
    return true
  }

  public isInvariant(): true {
    return true
  }

  public keys(): number[] {
    return []
  }
}

const NIL_NODE = new NilNode()

Object.freeze(NIL_NODE)

type Node<T> = NonNilNode<T> | NilNode

interface NodeConstructorArgs<T> {
  key: number
  value: T
  level: number
  left?: Node<T>
  right?: Node<T>
}

class UnreachableCaseError extends Error {
  constructor(val: never) {
    super(`Unreachable case: ${val}`)
  }
}

class NonNilNode<T> {
  public key: number
  public value: T
  public level: number
  public left: Node<T>
  public right: Node<T>

  public constructor({ key, value, level, left = NIL_NODE, right = NIL_NODE }: NodeConstructorArgs<T>) {
    this.key = key
    this.value = value
    this.level = level
    this.left = left
    this.right = right
  }

  public remove(key: number): Node<T> {
    const { left, right } = this

    if (key === this.key) {
      if (left.empty()) {
        return right
      } else if (right.empty()) {
        return left
      } else {
        const [lastKey, lastValue] = left.last()
        return this.clone({
          key: lastKey,
          value: lastValue,
          left: left.deleteLast(),
        }).adjust()
      }
    } else if (key < this.key) {
      return this.clone({
        left: left.remove(key),
      }).adjust()
    } else {
      return this.clone({
        right: right.remove(key),
      }).adjust()
    }
  }

  public empty(): this is NilNode {
    return false
  }

  public find(key: number): T | void {
    if (key === this.key) {
      return this.value
    } else if (key < this.key) {
      return this.left.find(key)
    } else {
      return this.right.find(key)
    }
  }

  public findWith(callback: FindCallback<T>): [number, T] | void {
    const result = callback(this.value)

    switch (result) {
      case -1:
        return this.left.findWith(callback)
      case 0:
        return [this.key, this.value]
      case 1:
        return this.right.findWith(callback)
      default:
        throw new UnreachableCaseError(result)
    }
  }

  public findMax(key: number): number {
    if (this.key === key) {
      return key
    }

    if (this.key < key) {
      const rightKey = this.right.findMax(key)
      if (rightKey === -Infinity) {
        return this.key
      } else {
        return rightKey
      }
    }

    return this.left.findMax(key)
  }

  public insert(key: number, value: T): NonNilNode<T> {
    if (key === this.key) {
      return this.clone({ key, value })
    } else if (key < this.key) {
      return this.clone({
        left: this.left.insert(key, value),
      }).rebalance()
    } else {
      return this.clone({
        right: this.right.insert(key, value),
      }).rebalance()
    }
  }

  public walkWithin(start: number, end: number): NodeData<T>[] {
    const { key, value } = this
    let result: NodeData<T>[] = []
    if (key > start) {
      result = result.concat(this.left.walkWithin(start, end))
    }

    if (key >= start && key <= end) {
      result.push({ key, value })
    }

    if (key <= end) {
      result = result.concat(this.right.walkWithin(start, end))
    }

    return result
  }

  public walk(): NodeData<T>[] {
    return [...this.left.walk<T>(), { key: this.key, value: this.value }, ...this.right.walk<T>()]
  }

  public last(): [number, T] {
    if (this.right.empty()) {
      return [this.key, this.value]
    } else {
      return this.right.last()
    }
  }

  public deleteLast(): Node<T> {
    if (this.right.empty()) {
      return this.left
    } else {
      return this.clone({
        right: this.right.deleteLast(),
      }).adjust()
    }
  }

  public clone(args: Partial<NodeConstructorArgs<T>>): NonNilNode<T> {
    return new NonNilNode<T>({
      key: args.key !== undefined ? args.key : this.key,
      value: args.value !== undefined ? args.value : this.value,
      level: args.level !== undefined ? args.level : this.level,
      left: args.left !== undefined ? args.left : this.left,
      right: args.right !== undefined ? args.right : this.right,
    })
  }

  public isSingle(): boolean {
    return this.level > this.right.level
  }

  public rebalance(): NonNilNode<T> {
    return this.skew().split()
  }

  public adjust(): NonNilNode<T> {
    const { left, right, level } = this
    if (right.level >= level - 1 && left.level >= level - 1) {
      return this
    } else if (level > right.level + 1) {
      if (left.isSingle()) {
        return this.clone({ level: level - 1 }).skew()
      } else {
        if (!left.empty() && !left.right.empty()) {
          return left.right.clone({
            left: left.clone({ right: left.right.left }),
            right: this.clone({
              left: left.right.right,
              level: level - 1,
            }),
            level: level,
          })
        } else {
          throw new Error('Unexpected empty nodes')
        }
      }
    } else {
      if (this.isSingle()) {
        return this.clone({ level: level - 1 }).split()
      } else {
        if (!right.empty() && !right.left.empty()) {
          const rl = right.left
          const rightLevel = rl.isSingle() ? right.level - 1 : right.level

          return rl.clone({
            left: this.clone({
              right: rl.left,
              level: level - 1,
            }),
            right: right.clone({ left: rl.right, level: rightLevel }).split(),
            level: rl.level + 1,
          })
        } else {
          throw new Error('Unexpected empty nodes')
        }
      }
    }
  }

  public isInvariant(): boolean {
    const { left, right, level } = this

    if (level !== left.level + 1) {
      return false
    } else if (level !== right.level && level !== right.level + 1) {
      return false
    } else if (!right.empty() && level <= right.right.level) {
      return false
    } else {
      return left.isInvariant() && right.isInvariant()
    }
  }

  public keys(): number[] {
    return [...this.left.keys(), this.key, ...this.right.keys()]
  }

  public ranges(): Range<T>[] {
    return this.toRanges(this.walk())
  }

  public rangesWithin(startIndex: number, endIndex: number): Range<T>[] {
    return this.toRanges(this.walkWithin(startIndex, endIndex))
  }

  private toRanges(nodes: NodeData<T>[]): Range<T>[] {
    if (nodes.length === 0) {
      return []
    }

    const first = nodes[0]

    let { key: start, value } = first

    const result = []

    for (let i = 1; i <= nodes.length; i++) {
      let nextNode = nodes[i]
      let end = nextNode ? nextNode.key - 1 : Infinity
      result.push({ start, end, value })

      if (nextNode) {
        start = nextNode.key
        value = nextNode.value
      }
    }
    return result
  }

  private split(): NonNilNode<T> {
    const { right, level } = this
    if (!right.empty() && !right.right.empty() && right.level == level && right.right.level == level) {
      return right.clone({
        left: this.clone({ right: right.left }),
        level: level + 1,
      })
    } else {
      return this
    }
  }

  private skew(): NonNilNode<T> {
    const { left } = this

    if (!left.empty() && left.level === this.level) {
      return left.clone({
        right: this.clone({ left: left.right }),
      })
    } else {
      return this
    }
  }
}

export class AATree<T> {
  private root: Node<T>

  public static empty<T>(): AATree<T> {
    return new AATree<T>(NIL_NODE)
  }

  private constructor(root: Node<T>) {
    this.root = root
  }

  public find(key: number): T | void {
    return this.root.find(key)
  }

  public findMax(key: number): number {
    return this.root.findMax(key)
  }

  public findWith(callback: FindCallback<T>): [number, T] | void {
    return this.root.findWith(callback)
  }

  public insert(key: number, value: T): AATree<T> {
    return new AATree(this.root.insert(key, value))
  }

  public remove(key: number): AATree<T> {
    return new AATree(this.root.remove(key))
  }

  public empty() {
    return this.root.empty()
  }

  public keys(): number[] {
    return this.root.keys()
  }

  public walk(): NodeData<T>[] {
    return this.root.walk()
  }

  public walkWithin(start: number, end: number): NodeData<T>[] {
    let adjustedStart = this.root.findMax(start)
    return this.root.walkWithin(adjustedStart, end)
  }

  public ranges(): Range<T>[] {
    return this.root.ranges()
  }

  public rangesWithin(start: number, end: number): Range<T>[] {
    let adjustedStart = this.root.findMax(start)
    return this.root.rangesWithin(adjustedStart, end)
  }

  public isInvariant(): boolean {
    return this.root.isInvariant()
  }
}
