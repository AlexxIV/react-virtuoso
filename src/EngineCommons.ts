import { subject, skip, mapTo, debounceTime, TObservable } from './tinyrx'

interface TScrollLocationWithAlign {
  index: number
  align: 'start' | 'center' | 'end'
  behavior?: 'smooth' | 'auto'
}

export type TScrollLocation = number | TScrollLocationWithAlign

export const buildIsScrolling = (scrollTop$: TObservable<number>): TObservable<boolean> => {
  const isScrolling$ = subject(false)

  scrollTop$.pipe(skip(1), mapTo(true)).subscribe(isScrolling$.next)

  scrollTop$.pipe(skip(1), mapTo(false), debounceTime(200)).subscribe(isScrolling$.next)

  return isScrolling$
}
