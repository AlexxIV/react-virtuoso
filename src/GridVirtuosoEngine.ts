import {
  subject,
  map,
  scan,
  /*
  withLatestFrom,
  debounceTime,
  mapTo,
  skip,
  filter,
  coldSubject,
  */
  combineLatest,
  TObservable,
} from '../src/tinyrx'
import { makeInput, makeOutput } from './rxio'

type GridDimensions = [
  number, // container width,
  number, // container height,
  number | undefined, // item width,
  number | undefined // item height
]

type GridItemRange = [
  number, // start index
  number // end index
]

// ---Input
// container size
// item size
// total items
// scroll top
// overscan
// ---Output
// totalHeight
// [startIndex, endIndex]
// listOffset

export const GridVirtuosoEngine = () => {
  const gridDimensions$ = subject<GridDimensions>()
  const totalCount$ = subject(0)
  const scrollTop$ = subject(0)
  const overscan$ = subject(0)

  const itemRange$: TObservable<GridItemRange> = combineLatest(
    gridDimensions$,
    scrollTop$,
    overscan$,
    totalCount$
  ).pipe(
    scan(
      (
        itemRange,
        [[containerWidth, containerHeight, itemWidth, itemHeight], scrollTop, overscan, totalCount]
      ): GridItemRange => {
        if (itemWidth === undefined || itemHeight === undefined) {
          return [0, 0] as GridItemRange
        }

        const [startIndex, endIndex] = itemRange

        const itemsPerRow = Math.floor(containerWidth / itemWidth)
        const rowCount = Math.ceil(containerHeight / itemHeight)
        const scrollOffset = Math.floor(scrollTop / itemHeight) * itemsPerRow

        const listTop = itemHeight * Math.floor(startIndex / itemsPerRow)

        if (listTop > scrollTop) {
          let startIndexWithOverscan = Math.floor((scrollTop - overscan * 2) / itemHeight) * itemsPerRow
          startIndexWithOverscan = Math.max(0, startIndexWithOverscan)
          return [startIndexWithOverscan, itemsPerRow * rowCount - 1 + scrollOffset] as GridItemRange
        }

        const listBottom = itemHeight * Math.floor(endIndex / itemsPerRow) + itemHeight

        // console.log({ listTop, listBottom, scrollTop, containerHeight })

        if (listBottom < scrollTop + containerHeight) {
          const rowsWithOverscan = Math.ceil((containerHeight + overscan * 2) / itemHeight)
          let endIndex = itemsPerRow * rowsWithOverscan - 1 + scrollOffset
          endIndex = Math.min(endIndex, totalCount - 1)
          return [scrollOffset, endIndex] as GridItemRange
        }

        return itemRange
      },
      [0, 0]
    )
  )

  const totalHeight$ = combineLatest(gridDimensions$, totalCount$).pipe(
    map(([[containerWidth, _, itemWidth, itemHeight], totalCount]) => {
      if (itemWidth === undefined || itemHeight === undefined) {
        return 0
      }
      const itemsPerRow = Math.floor(containerWidth / itemWidth)
      const rows = Math.ceil(totalCount / itemsPerRow)
      return itemHeight * rows
    })
  )

  return {
    gridDimensions: makeInput(gridDimensions$),
    totalCount: makeInput(totalCount$),
    scrollTop: makeInput(scrollTop$),
    overscan: makeInput(overscan$),

    itemRange: makeOutput(itemRange$),
    totalHeight: makeOutput(totalHeight$),
  }
}
