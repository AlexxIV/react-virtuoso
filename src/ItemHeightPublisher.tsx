import { CallbackRefParam } from './Utils'
import { Item } from './OffsetList'
import ResizeObserver from 'resize-observer-polyfill'
import { ItemHeight } from 'VirtuosoStore'
import { TInput } from 'rxio'

export class ItemHeightPublisher {
  private itemHeights: TInput<ItemHeight[]>
  private itemElements: HTMLElement[] = []
  public observer?: ResizeObserver

  public constructor(itemHeights: TInput<ItemHeight[]>) {
    this.itemHeights = itemHeights
  }

  private publishSizes(items: HTMLElement[]) {
    const results: ItemHeight[] = []
    for (const item of items) {
      const index = parseInt(item.dataset.index!)
      const size = item.offsetHeight
      if (results.length === 0 || results[results.length - 1].size !== size) {
        results.push({ start: index, end: index, size })
      } else {
        results[results.length - 1].end++
      }
    }
    this.itemHeights(results)
  }

  public init() {
    this.observer = new ResizeObserver((entries: ResizeObserverEntry[]) => {
      this.publishSizes(entries.map(({ target }) => target as HTMLElement))
    })
    this.publishSizes(this.itemElements)
    this.itemElements.map(item => {
      this.observer!.observe(item)
    })
  }

  public destroy() {
    this.itemElements = []
    this.observer!.disconnect()
  }

  public trackRef = (ref: CallbackRefParam) => {
    if (ref) {
      this.itemElements.push(ref)
    }
  }

  public getItemAttributes() {
    return (item: Item) => {
      return {
        'data-index': item.index,
        'data-known-size': item.size,
        ref: this.trackRef,
      }
    }
  }
}
