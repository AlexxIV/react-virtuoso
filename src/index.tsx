export * from './Virtuoso'
export * from './GroupedVirtuoso'
