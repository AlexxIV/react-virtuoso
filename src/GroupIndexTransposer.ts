import { AATree } from './AATree'
import { Item } from './OffsetList'

export interface RecordItem extends Item {
  type: 'item'
  transposedIndex: number
  groupIndex: number
}

export interface GroupItem extends Item {
  type: 'group'
  groupIndex: number
}

export type ListItem = RecordItem | GroupItem

export class StubIndexTransposer {
  public transpose(items: Item[]): RecordItem[] {
    return items.map(item => {
      return {
        groupIndex: 0,
        index: item.index,
        offset: item.offset,
        size: item.size,
        transposedIndex: item.index,
        type: 'item',
      }
    })
  }
}

export class GroupIndexTransposer {
  public tree: AATree<[number, number]>
  private count: number

  public constructor(counts: number[]) {
    this.count = counts.reduce((acc, groupCount) => acc + groupCount + 1, 0)
    let tree = AATree.empty<[number, number]>()
    let groupIndex = 0
    let total = 0
    for (const groupCount of counts) {
      tree = tree.insert(total, [groupIndex, total])
      groupIndex++
      total += groupCount + 1
    }
    this.tree = tree
  }

  public totalCount(): number {
    return this.count
  }

  public transpose(items: Item[]): ListItem[] {
    return items.map(item => {
      const groupMatch = this.tree.find(item.index)
      if (groupMatch) {
        return {
          groupIndex: groupMatch[0],
          index: item.index,
          offset: item.offset,
          size: item.size,
          type: 'group',
        }
      }

      const [groupIndex] = this.tree.findMaxValue(item.index)!
      return {
        groupIndex: groupIndex,
        index: item.index,
        offset: item.offset,
        size: item.size,
        transposedIndex: item.index - groupIndex - 1,
        type: 'item',
      }
    })
  }

  public groupIndices() {
    return this.tree.keys()
  }
}
