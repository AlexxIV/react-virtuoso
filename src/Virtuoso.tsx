import React, { CSSProperties, PureComponent, ReactElement, FC } from 'react'
import { VirtuosoContext } from './VirtuosoContext'
import { VirtuosoStore } from './VirtuosoStore'
import { VirtuosoView } from './VirtuosoView'
import { ListItem } from './GroupIndexTransposer'
import { TRender } from './VirtuosoList'

export type VirtuosoState = ReturnType<typeof VirtuosoStore>

export interface VirtuosoProps {
  totalCount: number
  overscan?: number
  topItems?: number
  footer?: () => ReactElement
  item: (index: number) => ReactElement
  itemHeight?: number
  endReached?: (index: number) => void
  scrollingStateChange?: (isScrolling: boolean) => void
  style?: CSSProperties
  className?: string
}

interface TVirtuosoPresentationProps {
  contextValue: VirtuosoState
  item: TRender
  footer?: () => ReactElement
  style?: CSSProperties
  className?: string
  itemHeight?: number
}

export const VirtuosoPresentation: FC<TVirtuosoPresentationProps> = ({
  contextValue,
  style,
  className,
  item,
  footer,
  itemHeight,
}) => {
  return (
    <VirtuosoContext.Provider value={contextValue}>
      <VirtuosoView
        style={style || {}}
        className={className}
        item={item}
        footer={footer}
        fixedItemHeight={itemHeight !== undefined}
      />
    </VirtuosoContext.Provider>
  )
}

export class Virtuoso extends PureComponent<VirtuosoProps, VirtuosoState> {
  public constructor(props: VirtuosoProps) {
    super(props)
    this.state = VirtuosoStore(props)
  }

  public static getDerivedStateFromProps(props: VirtuosoProps, state: VirtuosoState) {
    state.isScrolling(props.scrollingStateChange)
    state.endReached(props.endReached)
    state.topItemCount(props.topItems || 0)
    state.totalCount(props.totalCount)
    return null
  }

  private itemRenderer = (item: ListItem) => {
    return this.props.item(item.index)
  }

  public render() {
    return (
      <VirtuosoPresentation
        contextValue={this.state}
        style={this.props.style}
        className={this.props.className}
        item={this.itemRenderer}
        footer={this.props.footer}
        itemHeight={this.props.itemHeight}
      />
    )
  }
}
