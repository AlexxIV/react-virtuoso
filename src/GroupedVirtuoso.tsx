import React, { forwardRef, ReactElement, useImperativeHandle, useState, useEffect } from 'react'
import { TScrollLocation } from './EngineCommons'
import { TItemContainer, VirtuosoPresentation, VirtuosoProps } from './Virtuoso'
import { VirtuosoStore } from './VirtuosoStore'

type GroupedVirtuosoProps = Pick<VirtuosoProps, Exclude<keyof VirtuosoProps, 'totalCount' | 'topItems' | 'item'>> & {
  groupCounts: number[]
  group: (groupIndex: number) => ReactElement
  item: (index: number, groupIndex?: number) => ReactElement
  groupIndices?: (indices: number[]) => void
  GroupContainer?: TItemContainer
}

export interface GroupedVirtuosoMethods {
  scrollToIndex(location: TScrollLocation): void
}

export const GroupedVirtuoso = forwardRef<GroupedVirtuosoMethods, GroupedVirtuosoProps>((props, ref) => {
  const [state] = useState(VirtuosoStore(props))
  useImperativeHandle(
    ref,
    () => ({
      scrollToIndex: (location: TScrollLocation) => {
        state.scrollToIndex(location)
      },
    }),
    [state]
  )

  useEffect(() => {
    state.endReached(props.endReached)
    state.rangeChanged(props.rangeChanged)
    state.atBottomStateChange(props.atBottomStateChange)
    state.isScrolling(props.scrollingStateChange)
    state.groupCounts(props.groupCounts)
    state.groupIndices(props.groupIndices)
    state.itemsRendered(props.itemsRendered)
    state.totalListHeightChanged(props.totalListHeightChanged)
    state.renderProp(props.item)
    state.groupRenderProp(props.group)
    state.itemContainer(props.ItemContainer || 'div')
    state.groupContainer(props.GroupContainer || 'div')

    return () => {
      state.itemsRendered(undefined)
      state.totalListHeightChanged(undefined)
    }
  }, [
    state,
    props.endReached,
    props.rangeChanged,
    props.atBottomStateChange,
    props.scrollingStateChange,
    props.groupCounts,
    props.groupIndices,
    props.itemsRendered,
    props.totalListHeightChanged,
    props.item,
    props.group,
    props.GroupContainer,
    props.ItemContainer,
  ])

  return (
    <VirtuosoPresentation
      contextValue={state}
      style={props.style}
      className={props.className}
      footer={props.footer}
      itemHeight={props.itemHeight}
      ScrollContainer={props.ScrollContainer}
      FooterContainer={props.FooterContainer}
      ListContainer={props.ListContainer}
    />
  )
})

GroupedVirtuoso.displayName = 'GroupedVirtuoso'
