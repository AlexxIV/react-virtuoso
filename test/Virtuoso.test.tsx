import React from 'react'
import renderer from 'react-test-renderer'
import { GroupedVirtuoso } from '../src'
import { Virtuoso } from '../src/Virtuoso'

test('Test rendering of empty list', () => {
  expect(
    renderer.create(
      <div>
        <Virtuoso totalCount={100} item={() => <div />} />
      </div>
    )
  ).toMatchSnapshot()
})

test('Test rendering of empty grouped list', () => {
  expect(
    renderer.create(
      <div>
        <GroupedVirtuoso groupCounts={[]} group={() => <div />} item={() => <div />} />
      </div>
    )
  ).toMatchSnapshot()
})
