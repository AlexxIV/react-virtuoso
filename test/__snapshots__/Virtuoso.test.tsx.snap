// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test rendering of empty grouped list 1`] = `
<div>
  <div
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "height": "40rem",
        "outline": "none",
        "overflowY": "auto",
        "position": "relative",
      }
    }
    tabIndex={0}
  >
    <div
      style={
        Object {
          "height": "100%",
          "overflow": "absolute",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "marginTop": "NaNpx",
          }
        }
      >
         
         
      </div>
    </div>
    <div
      style={
        Object {
          "height": "0px",
          "position": "absolute",
          "top": 0,
        }
      }
    >
       
    </div>
  </div>
</div>
`;

exports[`Test rendering of empty list 1`] = `
<div>
  <div
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "height": "40rem",
        "outline": "none",
        "overflowY": "auto",
        "position": "relative",
      }
    }
    tabIndex={0}
  >
    <div
      style={
        Object {
          "height": "100%",
          "overflow": "absolute",
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "marginTop": "NaNpx",
          }
        }
      >
         
         
      </div>
    </div>
    <div
      style={
        Object {
          "height": "0px",
          "position": "absolute",
          "top": 0,
        }
      }
    >
       
    </div>
  </div>
</div>
`;
